### PR TITLE
SMB for Cyberduck

### PR DESCRIPTION
**Pull request by mbscit**

### CODE REVIEW

File: smb/src/main/java/ch/cyberduck/core/smb/SMBListService.java [hunk 1]
```diff
@@ -76,7 +84,11 @@ public AttributedList<Path> list(final Path directory, final ListProgressListene
             return result;
         }
         catch(SMBApiException e) {
-            throw new SMBExceptionMappingService().map(e);
+            if(e.getStatusCode() == 3221226071L) {
+                throw new UnsupportedException("This folder is a DFS Share, which is currently not supported");
```
Comment (dkocher) [inline smb/src/main/java/ch/cyberduck/core/smb/SMBListService.java:88]: Does support depend on hierynomus/smbj#152?

File: smb/src/main/java/ch/cyberduck/core/smb/SMBListService.java [hunk 2]
```diff
@@ -76,7 +84,11 @@ public AttributedList<Path> list(final Path directory, final ListProgressListene
             return result;
         }
         catch(SMBApiException e) {
-            throw new SMBExceptionMappingService().map(e);
+            if(e.getStatusCode() == 3221226071L) {
```
Comment (dkocher) [inline smb/src/main/java/ch/cyberduck/core/smb/SMBListService.java:87]: Remove magic number and reference constant `STATUS_PATH_NOT_COVERED`.

File: smb/src/main/java/ch/cyberduck/core/smb/SMBListService.java (hunk 1 above)
Comment (dkocher) [inline smb/src/main/java/ch/cyberduck/core/smb/SMBListService.java:88]: Handle this error case in `SMBExceptionMappingService`.

File: smb/src/main/java/ch/cyberduck/core/smb/SMBTimestampFeature.java [hunk 3]
```diff
@@ -51,9 +51,7 @@ public void setTimestamp(Path file, TransferStatus status) throws BackgroundExce
             fileEntry.setFileInformation(fileBasicInformation);
         }
         catch(SMBApiException e) {
-            throw new SMBExceptionMappingService().map("Setting timestamp on file {0} failed", e, file);
-        }
-
+            throw new SMBExceptionMappingService().map("Cannot change timestamp of {0}", e, file);        }
```
Comment (dkocher) [inline smb/src/main/java/ch/cyberduck/core/smb/SMBTimestampFeature.java:54]: Run formatting.

File: smb/src/main/java/ch/cyberduck/core/smb/SMBListService.java [hunk 4]
```diff
@@ -40,9 +41,16 @@ public SMBListService(SMBSession session) {
     @Override
     public AttributedList<Path> list(final Path directory, final ListProgressListener listener) throws BackgroundException {
 
+        String dst = directory.getAbsolute();
+        if(dst.charAt(0) == '/') {
```
Comment (dkocher) [inline smb/src/main/java/ch/cyberduck/core/smb/SMBListService.java:45]: Should we have generic handling of this path conversion for all features that work with absolute paths from `Path`?

**Review by dkocher — CHANGES_REQUESTED** [review]

File: smb/src/main/java/ch/cyberduck/core/smb/SMBListService.java (hunk 1 above)
Comment (mbscit) [reply smb/src/main/java/ch/cyberduck/core/smb/SMBListService.java]: yes

File: smb/src/main/java/ch/cyberduck/core/smb/SMBListService.java (hunk 4 above)
Comment (mbscit) [reply smb/src/main/java/ch/cyberduck/core/smb/SMBListService.java]: The other calls seem to work with both formats.
But to make it the same everywhere, we have now added the class [SMBUtils](https://github.com/mbscit/smb-cyberduck/blob/feature/final-clean/smb/src/main/java/ch/cyberduck/core/smb/SMBUtils.java) that does this conversion, and applied it everywhere the absolute path is used.

File: smb/src/main/java/ch/cyberduck/core/smb/SMBCopyFeature.java [hunk 5]
```diff
@@ -34,28 +35,26 @@ public Path copy(Path source, Path target, TransferStatus status, ConnectionCall
...
             }
             else {
-                copyDirectory(source, target, status, prompt, listener);
+                copyDirectory(target);
             }
             return target;
         }
         catch(SMBApiException e) {
-            throw new SMBExceptionMappingService().map(e);
+            throw new BackgroundException(MessageFormat.format("Cannot copy {0}", source), e);
```
Comment (dkocher) [inline smb/src/main/java/ch/cyberduck/core/smb/SMBCopyFeature.java:46]:
```suggestion
            throw new SMBExceptionMappingService().map("Cannot copy {0}", e, source);
```

File: smb/src/main/java/ch/cyberduck/core/smb/SMBUtils.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
+package ch.cyberduck.core.smb;/*
```
Comment (dkocher) [inline smb/src/main/java/ch/cyberduck/core/smb/SMBUtils.java:1]: Format and remove duplicate copyright header.

**Review by dkocher — CHANGES_REQUESTED** [review]

**Comment by dkocher** [comment]
Replaced with #15009.

**Pull request closed without merging**